packages/itwinui-react/src/core/utils/components/ShadowRoot.tsx

```diff
@@ -68,19 +38,90 @@ export const ShadowRoot = ({
     return null;
   }
 
-  return (
-    <>
-      {shadowRoot ? (
-        ReactDOM.createPortal(children, shadowRoot)
-      ) : (
-        <template ref={attachShadowRef} />
-      )}
-    </>
+  return <ClientShadowRoot css={css}>{children}</ClientShadowRoot>;
+};
+
+// ----------------------------------------------------------------------------
+
+const ClientShadowRoot = ({ children, css }: ShadowRootProps) => {
+  const templateRef = React.useRef<HTMLTemplateElement>(null);
+  const shadowRoot = useShadowRoot(templateRef, { css });
+
+  // fallback to <style> tag if adoptedStyleSheets is not supported
+  const fallbackCss =
+    !supportsAdoptedStylesheets && css ? <style>{css}</style> : null;
+
+  return shadowRoot ? (
+    ReactDOM.createPortal(
+      <>
+        {fallbackCss}
+        {children}
+      </>,
+      shadowRoot,
+    )
+  ) : (
+    <template ref={templateRef} />
   );
 };
 
 // ----------------------------------------------------------------------------
 
+/**
+ * Given a ref, this hook will return a shadowRoot attached to its parent element.
+ *
+ * The css will be added to the shadowRoot using `adoptedStyleSheets` (if supported).
+ */
+function useShadowRoot(
+  templateRef: React.RefObject<HTMLElement>,
+  { css = '' },
+) {
+  const [shadowRoot, setShadowRoot] = React.useState<ShadowRoot | null>(null);
+  const styleSheet = React.useRef<CSSStyleSheet>();
+  const latestCss = useLatestRef(css);
+
+  useLayoutEffect(() => {
+    const parent = templateRef.current?.parentElement;
+    if (!parent) {
+      return;
+    }
+
+    if (parent.shadowRoot) {
+      parent.shadowRoot.replaceChildren(); // Remove previous shadowroot content
+    }
+
+    const shadow = parent.shadowRoot || parent.attachShadow({ mode: 'open' });
+
+    if (supportsAdoptedStylesheets) {
+      // create an empty stylesheet and add it to the shadowRoot
+      const _global = shadow.ownerDocument.defaultView || globalThis;
+      styleSheet.current = new _global.CSSStyleSheet();
+      shadow.adoptedStyleSheets = [styleSheet.current];
+
+      // add the CSS immediately to avoid FOUC (one-time)
+      if (latestCss.current) {
+        styleSheet.current.replaceSync(latestCss.current);
+      }
+    }
+
+    queueMicrotask(() => {
+      ReactDOM.flushSync(() => setShadowRoot(shadow));
+    });
+
+    return () => void setShadowRoot(null);
+  }, [templateRef, latestCss]);
+
+  // Synchronize the CSS contents of existing stylesheet
+  useLayoutEffect(() => {
+    if (css && supportsAdoptedStylesheets) {
+      styleSheet.current?.replaceSync(css);
+    }
+  }, [css]);
```

Comment:
Since `latestCss` is listed as a dependency in the above `useLayoutEffect`, will the above `useLayoutEffect` handle what this `useLayoutEffect` is trying to handle?

Reply:
nope, it's a ref